src/sentry/snuba/discover.py

```diff
@@ -611,7 +612,66 @@ def query(
     return transform_results(result, translated_columns, snuba_args)
 
 
-def timeseries_query(selected_columns, query, params, rollup, reference_event=None, referrer=None):
+def key_transaction_conditions(queryset):
+    """
+        The snuba query for transactions is of the form
+        (transaction="1" AND project=1) OR (transaction="2" and project=2) ...
+        which the schema intentionally doesn't support so we cannot do an AND in OR
+        so here the "and" operator is being instead to do an AND in OR query
+    """
+    return [
+        [
+            # First layer is Ands
+            [
+                # Second layer is Ors
+                [
+                    "and",
+                    [
+                        [
+                            "equals",
+                            # Without the outer ' here, the transaction will be treated as another column
+                            # instead of a string. This isn't an injection risk since snuba is smart enough to
+                            # handle escaping for us.
+                            ["transaction", u"'{}'".format(transaction.transaction)],
+                        ],
+                        ["equals", ["project_id", transaction.project.id]],
+                    ],
+                ],
+                "=",
+                1,
+            ]
+            for transaction in queryset
+        ]
+    ]
+
+
+def key_transaction_query(selected_columns, user_query, params, orderby, referrer, queryset):
+    return query(
+        selected_columns,
+        user_query,
+        params,
+        orderby=orderby,
+        referrer=referrer,
+        conditions=key_transaction_conditions(queryset),
+    )
+
+
+def key_transaction_timeseries_query(
+    selected_columns, user_query, params, rollup, referrer, queryset
+):
+    return timeseries_query(
+        selected_columns,
+        user_query,
+        params,
+        rollup,
+        referrer=referrer,
```

Comment:
It would be good if key transaction queries had a distinct referrer so we can generate aggregates on them in our internal discover reports.

Reply:
:+1: Referrer is being set from the API side, but should I move that here instead?